client/fetch_tagged_results_accumulator_merge_test.go

```diff
@@ -0,0 +1,581 @@
+	require.NoError(t, err)
+	require.False(t, results.Exhaustive)
+	matcher := newTestSerieses(1, 8).indexMatcher()
+	require.True(t, matcher.Matches(results.Iterator))
+
+	iters, exhaust, err := accum.AsEncodingSeriesIterators(10, th.pools)
+	require.NoError(t, err)
+	require.False(t, exhaust)
+	append(sg0, sg1...).assertMatchesEncodingIters(t, iters)
+}
+
+func TestFetchTaggedResultsAccumulatorSeriesItersDatapointsNSplit(t *testing.T) {
+	// rf=3, 3 identical hosts, with same shards
+	topoMap := testutil.MustNewTopologyMap(3, map[string][]shard.Shard{
+		"testhost0": testutil.ShardsRange(0, 29, shard.Available),
+		"testhost1": testutil.ShardsRange(0, 29, shard.Available),
+		"testhost2": testutil.ShardsRange(0, 29, shard.Available),
+	})
+
+	var (
+		sg0       = newTestSerieses(1, 10)
+		startTime = time.Now().Add(-time.Hour).Truncate(time.Hour)
+		endTime   = time.Now().Truncate(time.Hour)
+		numPoints = 100
+	)
+	sg0.addDatapoints(numPoints, startTime, endTime)
+	groups := sg0.nsplit(3)
+
+	th := newTestFetchTaggedHelper(t)
+	workflow := testFetchTaggedWorkflow{
+		t:         t,
+		topoMap:   topoMap,
+		level:     topology.ReadConsistencyLevelAll,
+		startTime: startTime,
+		endTime:   endTime,
+		steps: []testFetchTaggedWorklowStep{
+			testFetchTaggedWorklowStep{
+				hostname: "testhost0",
+				response: groups[0].toRPCResult(th, startTime, true),
+			},
+			testFetchTaggedWorklowStep{
+				hostname: "testhost1",
+				response: groups[1].toRPCResult(th, endTime, true),
+			},
+			testFetchTaggedWorklowStep{
+				hostname:     "testhost2",
+				response:     groups[2].toRPCResult(th, endTime, true),
+				expectedDone: true,
+			},
+		},
+	}
+	accum := workflow.run()
+
+	results, err := accum.AsIndexQueryResults(8, th.pools)
+	require.NoError(t, err)
+	require.False(t, results.Exhaustive)
+	matcher := newTestSerieses(1, 8).indexMatcher()
+	require.True(t, matcher.Matches(results.Iterator))
+
+	iters, exhaust, err := accum.AsEncodingSeriesIterators(10, th.pools)
+	require.NoError(t, err)
+	require.True(t, exhaust)
+	// ensure iters are valid after the lifecycle of the accumulator
+	accum.Clear()
+	sg0.assertMatchesEncodingIters(t, iters)
+}
+
+// debugIter is useful for tests, leaving in for now.
+// nolint
+type debugIter struct {
+	index.Iterator
+}
+
+// nolint
+func (d debugIter) String() string {
+	iter := d.Iterator
+	var buffer bytes.Buffer
+	for iter.Next() {
+		ns, id, tags := iter.Current()
+		buffer.WriteString(fmt.Sprintf("ns: %v\n", ns.String()))
+		buffer.WriteString(fmt.Sprintf("id: %v\n", id.String()))
+		for tags.Next() {
+			t := tags.Current()
+			buffer.WriteString(fmt.Sprintf("tag: [ name = %v, value = %v ]\n", t.Name.String(), t.Value.String()))
+		}
+		if err := tags.Err(); err != nil {
+			buffer.WriteString(fmt.Sprintf("tag-err: %v\n", err))
+		}
+	}
+	if err := iter.Err(); err != nil {
+		buffer.WriteString(fmt.Sprintf("err: %v\n", err))
+	}
+	return buffer.String()
+}
+
+type testFetchTaggedWorkflow struct {
```

Comment:
We should definitely write more declarative tests like this.  There's a few, but not many that are as well defined like this. 👍

Reply:
For sure. They capture intent much more succinctly, will standup better to code evolution.